lib/sync/fullsync.go

```diff
@@ -0,0 +1,547 @@
+			if err != nil {
+				return false, nil, nil, fmt.Errorf("while handling ready block: %w", err)
+			}
+
+			if imported {
+				f.syncedBlocks += 1
+			}
+		}
+
+		nextBlocksToImport = make([]*types.BlockData, 0)
+
+		// check if blocks from the disjoint set can be imported on their on forks
+		// given that fragment contains chains and these chains contains blocks
+		// check if the first block in the chain contains a parent known by us
+		for _, fragment := range disjointFragments {
+			highestFinalized, err := f.blockState.GetHighestFinalisedHeader()
+			if err != nil {
+				return false, nil, nil, fmt.Errorf("getting highest finalized header")
+			}
+
+			validFragment := validBlocksUnderFragment(highestFinalized.Number, fragment)
+			if len(validFragment) == 0 {
+				continue
+			}
+
+			ok, err := f.blockState.HasHeader(validFragment[0].Header.ParentHash)
+			if err != nil && !errors.Is(err, database.ErrNotFound) {
+				return false, nil, nil, err
+			}
+
+			if !ok {
+				// if the parent of this valid fragment is behind our latest finalized number
+				// then we can discard the whole fragment since it is a invalid fork
+				if (validFragment[0].Header.Number - 1) <= highestFinalized.Number {
+					continue
+				}
+
+				logger.Infof("starting an acestor search from %s parent of #%d (%s)",
+					validFragment[0].Header.ParentHash,
+					validFragment[0].Header.Number,
+					validFragment[0].Header.Hash(),
+				)
+
+				f.unreadyBlocks.newFragment(validFragment)
+				request := network.NewBlockRequest(
+					*variadic.FromHash(validFragment[0].Header.ParentHash),
+					network.MaxBlocksInResponse,
+					network.BootstrapRequestData, network.Descending)
+				f.requestQueue.PushBack(request)
+			} else {
+				// inserting them in the queue to be processed after the main chain
+				nextBlocksToImport = append(nextBlocksToImport, validFragment...)
+			}
+		}
+
+		disjointFragments = nil
+	}
+
+	return false, repChanges, peersToIgnore, nil
+}
+
+func (f *FullSyncStrategy) ShowMetrics() {
+	totalSyncAndImportSeconds := time.Since(f.startedAt).Seconds()
+	bps := float64(f.syncedBlocks) / totalSyncAndImportSeconds
+	logger.Infof("⛓️ synced %d blocks, disjoint fragments %d, incomplete blocks %d, "+
+		"took: %.2f seconds, bps: %.2f blocks/second, target block number #%d",
+		f.syncedBlocks, len(f.unreadyBlocks.disjointChains), len(f.unreadyBlocks.incompleteBlocks),
+		totalSyncAndImportSeconds, bps, f.peers.getTarget())
+}
+
+func (f *FullSyncStrategy) OnBlockAnnounceHandshake(from peer.ID, msg *network.BlockAnnounceHandshake) error {
+	f.peers.update(from, msg.BestBlockHash, msg.BestBlockNumber)
+	return nil
+}
+
+func (f *FullSyncStrategy) OnBlockAnnounce(from peer.ID, msg *network.BlockAnnounceMessage) (repChange *Change, err error) {
+	if f.blockState.IsPaused() {
+		return nil, errors.New("blockstate service is paused")
+	}
+
+	if msg.BestBlock {
+		pv := f.peers.get(from)
```

Comment:
I'm confused what is happening here, why is this bad behavior? Is there a change that our peer view could become outdated and thus not equal the messages best block? Maybe I am misunderstanding what is occurring

Reply:
when a peer announces a block it first sends a handshaking with his best block hash and number (and genesis hash), we store that handshake and then it sends the block announce, in the block announce message it can set if the block being announce is his best block, in such case I put this validation to check if the block sent in the handshake as the best block matches with the announced if `msg.BestBlock == true`